lib/plugins/aws/invokeLocal/index.js

```diff
@@ -104,53 +124,183 @@ class AwsInvokeLocal {
+
+      // profile override from config
+      const profileOverride = this.provider.getProfile();
+      if (profileOverride) {
+        lambdaDefaultEnvVars.AWS_PROFILE = profileOverride;
       }
-    }
 
-    // profile override from config
-    const profileOverride = this.provider.getProfile();
-    if (profileOverride) {
-      lambdaDefaultEnvVars.AWS_PROFILE = profileOverride;
+      const configuredEnvVars = this.getConfiguredEnvVars();
+
+      const promises = _.entries(configuredEnvVars).map(([name, value]) => {
+        return this.resolveResource(value).then(output => {
+          configuredEnvVars[name] = output;
+        });
+      });
+
+      return BbPromise.all(promises).then(() => {
+        Object.assign(process.env, lambdaDefaultEnvVars, credentialEnvVars, configuredEnvVars);
+      });
+    });
+  }
+
+  resolveExports() {
+    return this.resolveExportsImpl();
+  }
+
+  resolveExportsImpl(token = undefined, cfExports = undefined) {
+    if (!cfExports) {
+      cfExports = [];
     }
+    return this.provider
+      .request('CloudFormation', 'listExports', { NextToken: token })
+      .then(result => {
+        cfExports.push(...result.Exports);
+        if (result.NextToken) {
+          return this.resolveExportsImpl(result.NextToken, cfExports);
+        }
 
-    const configuredEnvVars = this.getConfiguredEnvVars();
+        return cfExports;
+      });
+  }
 
-    _.merge(process.env, lambdaDefaultEnvVars, credentialEnvVars, configuredEnvVars);
+  resolveResource(value) {
+    return BbPromise.try(() => {
+      if (!_.isObject(value)) {
+        return value;
+      }
+      let resourceName = null;
+      let action = null;
+
+      if (value.Ref) {
+        action = 'Ref';
+        resourceName = value[action];
+      } else if (value['Fn::GetAtt']) {
+        action = 'Fn::GetAtt';
+
+        // Handle condition where GetAtt has Resource.Property format
+        if (typeof value[action] === 'string') {
+          value[action] = value[action].split('.');
+        }
+        if (value[action].length !== 2) {
+          throw new ServerlessError(`Attribute definition unexpected ${JSON.stringify(value)}`);
+        }
 
-    return BbPromise.resolve();
+        // Get the resource name
+        resourceName = value[action][0];
+      } else if (value['Fn::ImportValue']) {
+        action = 'Fn::ImportValue';
+        resourceName = value[action];
+      } else {
+        return BbPromise.resolve(value);
+      }
+
+      let resourceResult;
+      if (action === 'Fn::ImportValue') {
+        resourceResult = this.resolveExports().then(cfExports => {
+          const result = cfExports.find(exp => exp.Name === resourceName);
+          if (!result) {
+            throw new ServerlessError(`Could not find import value for ${resourceName}`);
+          }
+          return result.Value;
+        });
+      } else {
+        resourceResult = this.getCFResources().then(cfResources => {
```

Comment:
We should refer to resources that reflect the local state of configuration. What was deployed might be not in sync with that

So I believe we should ensure that full CF template compilation is made and then inspect `this.provider.compiledCloudFormationTemplate.resources` for needed generated resources

Reply:
But that goes back to your earlier argument that if the resource is not deployed we would have the person debugging dealing with issues interacting with those resources, and generating errors in configuring the environment since we're interogating cloudformation and aws anyway for resource values anyway.  This seems like you're trying to get at cross purposes, and it would be confusing the person using this.

Either the resources are deployed and they can be interacted with, or they're not and resource resolution is going to fail anyway.  At this point, there's no reason to resolve a resource that is not deployed when debugging and using that resource to interact with.

Reply:
I'm not sure if you understood me correctly. 

I believe that for local invocation we should test state as is in code and not as it's deployed (and current implementation as in `master` follows that).

Resolution of CloudFormation instructions in many cases can be done offline purely by inspecting the generated CloudFormation stack.

My point is, that we should inspect the generated CF stack (as to be deployed), and not internal serverless instance configuration, which is hard to cover fully, is subject to changes, and does not cover eventual changes as introduced by plugins.

That can't be done without regenerating template, which we can achieve that same way as when doing local invocation with docker, by spawning package command as here: https://github.com/serverless/serverless/blob/1793cf8d7a55b85fc6505ae493dcca2292e443d2/lib/plugins/aws/invokeLocal/index.js#L379

(Still as it'll prolong command execution, we should do it only if we're sure we need to inspect CloudFormation stack.)

Aside of above, there are cases when we're forced to rely purely on deployed state (e.g. when env var points some id that's generated on AWS side at the time of resource creation),, and yes, we should do that in such case, there's no other way. 

That's how I see this. Do you think its bad idea? Did I miss something?